dot/digest/digest.go

```diff
@@ -179,15 +189,41 @@ func (h *Handler) handleGrandpaConsensusDigest(digest scale.VaryingDataType, hea
 }
 
 func (h *Handler) handleBabeConsensusDigest(digest scale.VaryingDataType, header *types.Header) error {
+	h.babeConsensusDigestLock.Lock()
+	defer h.babeConsensusDigestLock.Unlock()
+
+	currEpoch, err := h.epochState.GetEpochForBlock(header)
+	if err != nil {
+		return err
+	}
+
+	nextEpoch := currEpoch + 1
+	headerHash := header.Hash()
+
 	switch val := digest.Value().(type) {
 	case types.NextEpochData:
-		h.logger.Debugf("handling BABENextEpochData data: %v", digest)
-		return h.handleNextEpochData(val, header)
+		h.logger.Debugf("stored BABENextEpochData data: %v for hash: %s to epoch: %d\n", digest, headerHash, nextEpoch)
+
+		_, has := h.nextEpochData[nextEpoch]
+		if !has {
+			h.nextEpochData[nextEpoch] = make(map[common.Hash]types.NextEpochData)
+		}
+		h.nextEpochData[nextEpoch][headerHash] = val
+
+		return h.handleNextEpochData(val, nextEpoch, header)
```

Comment:
this works for bootstrap mode, since there's only really 1 chain, but it doesn't entirely fix the problem in the case that the previous epoch hasn't been finalized yet. for example if we have two forks A and B, and B arrives later than A, so the epoch data for B will be set in the database. however if we process blocks on fork A, they will use the info from fork B, which will result in them being marked invalid and thrown away even if they are valid. I think you probably need to modify the babe verifier to get the data from the right fork, if it isn't set in the database already (and only set it in the database on finalization)